### ISSUE TITLE
[feature]: Add NSFW/TW filter to images

### ISSUE THREAD

**Issue by darigovresearch** [freesewing/freesewing#6005]
### What is it that you would like to see happen? 🤔

Would be great to have an NSFW filter for things like showcases of underwear being modelled etc like on Discord

**Comment by BenJamesBen** [comment]
2 proofs of concept:
1. `<Spoiler1>` uses a simple class that blurs images until hovered/focused.
2. `<Spoiler2>` uses checkboxes and discrete ids to blur images until clicked. Limit of 9 spoilers per page.

There might be better ways to do it, but I am not an expert in CSS.

```diff
--- a/sites/shared/components/mdx/index.mjs
+++ b/sites/shared/components/mdx/index.mjs
@@ -2,6 +2,8 @@
 import { Popout } from 'shared/components/popout/index.mjs'
 import { Highlight } from './highlight.mjs'
 import { YouTube } from './youtube.mjs'
+import { Spoiler1 } from './spoiler1.mjs'
+import { Spoiler2 } from './spoiler2.mjs'
 //import { Figure } from './figure.mjs'
 import { ReadMore } from './read-more.mjs'
 import { Tab, Tabs } from '../tabs.mjs'
@@ -34,6 +36,8 @@ export const components = (site = 'org', slug = []) => {
   const extra = {
     pre: (props) => <Highlight {...props} />,
     YouTube,
+    Spoiler1,
+    Spoiler2,
     // This Figure component causes hydration errors
     //img: Figure,
     table: (props) => (
```

```diff
--- /dev/null
+++ b/sites/shared/components/mdx/spoiler1.mjs
@@ -0,0 +1,3 @@
+export const Spoiler1 = (props) => {
+  return <img class="spoiler" src={props.src} alt={props.alt} />
+}
```

```diff
--- /dev/null
+++ b/sites/shared/components/mdx/spoiler2.mjs
@@ -0,0 +1,10 @@
+export const Spoiler2 = (props) => {
+  return (
+    <div>
+      <input type="checkbox" id={props.id} />
+      <label for={props.id}>
+        <img src={props.src} alt={props.alt} />
+      </label>
+    </div>
+  )
+}
```

```diff
--- a/sites/shared/styles/globals.css
+++ b/sites/shared/styles/globals.css
@@ -533,6 +533,69 @@ progress.progress-white::-moz-progress-bar {
   max-height: calc(100vh - 6rem);
 }
 
+/* NSFW/spoiler image blur */
+.spoiler {
+  filter: blur(20px);
+}
+.spoiler:hover,
+.spoiler:focus {
+  filter: blur(0px);
+}
+#spoiler1 + label > img {
+  filter: blur(20px);
+}
+#spoiler1:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler2 + label > img {
+  filter: blur(20px);
+}
+#spoiler2:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler3 + label > img {
+  filter: blur(20px);
+}
+#spoiler3:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler4 + label > img {
+  filter: blur(20px);
+}
+#spoiler4:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler5 + label > img {
+  filter: blur(20px);
+}
+#spoiler5:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler6 + label > img {
+  filter: blur(20px);
+}
+#spoiler6:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler7 + label > img {
+  filter: blur(20px);
+}
+#spoiler7:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler8 + label > img {
+  filter: blur(20px);
+}
+#spoiler8:checked + label > img {
+  filter: blur(0px);
+}
+#spoiler9 + label > img {
+  filter: blur(20px);
+}
+#spoiler9:checked + label > img {
+  filter: blur(0px);
+}
+
 /* Style for navigation */
 details {
   user-select: none;
```

![Screenshot 2024-02-17 at 6 33 54 AM](https://github.com/freesewing/freesewing/assets/109869956/cd2f9f0d-aa35-4ab1-9440-0b9e96780b01)

```diff
--- a/markdown/org/showcase/two-bruce-boxer-briefs/en.md
+++ b/markdown/org/showcase/two-bruce-boxer-briefs/en.md
@@ -7,7 +7,18 @@ title: "Two Bruce boxer briefs"
 designs: ["bruce"]
 ---
 
-The maker of these Bruce boxer briefs says they will be making even more pairs!
+The `<Spoiler1>` method (blur disappears when hovered/focused):
+<Spoiler1 src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-two-bruce-boxer-briefs/public" alt="ttttt" />
+<Spoiler1 src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-who-can-resist-a-cuddly-shark-named-hi/public" alt="ssssttttt" />
+<Spoiler1 src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-a-lined-tristan-top-with-front-lacing/public" alt="alt text" />
 
-![IMG_20230331_200738.jpg](https://posts.freesewing.org/uploads/IMG_20230331_200738_27f3f7b9bf.jpg)
+<br />
+<hr />
+<br />
+
+The `<Spoiler2>` method (blur disappears when clicked):
+
+<Spoiler2 id="spoiler1" src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-two-bruce-boxer-briefs/public" alt="ttttt" />
+<Spoiler2 id="spoiler2" src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-who-can-resist-a-cuddly-shark-named-hi/public" alt="ssssttttt" />
+<Spoiler2 id="spoiler3" src="https://imagedelivery.net/ouSuR9yY1bHt-fuAokSA5Q/showcase-a-lined-tristan-top-with-front-lacing/public" alt="alt text" />
 
```

**Comment by woutervdub** [comment]
I like this.

**Comment by darigovresearch** [comment]
Not sure, how do other tools that people use handle these guidelines?

**Comment by HaasJona** [comment]
I don't think you need the discrete IDs in the second case. Just put the checkbox inside the `<label>` and get rid of the ids.

**Comment by joostdecock** [comment]
Before we dive into how to implement this, I'd like to take a step back and talk about **why** we would implement this.

The technical implementation is not really a problem. The issue is that this will require moderation, and will probably open the door to endless discussions of what should or should not be blurred/hidden.

So, I would like to hear from the people who have been working on this whether that was because they thing it was important to have this, or rather because they want to help in any way they can and this issue was open, so why not?

As far as the implementation go, putting things in globals.css is an escape hatch. We use tailwind CSS so inline classes should do the trick unless there's no obvious way to do so.

Also note that an MDX component won't work for things like the main showcase image, which is probably one of the more important use-cases.

In summary:

- I am not convinced we want this (I am willing to listen to arguments in favor of it though)
- If we implement this, React state + inline classes is probably the way to go

**Comment by woutervdub** [comment]
After we discussed this some more (and I happily forgot where we did this) I would like to retract my opinion above, and change it to 'We should not do this'. Mainly for the reasons you just outlined (moderation). I'm also becoming less convinced that this is something that needs to be addressed. A sewing site with patterns seem to me to imply that there may be photos of said patterns. If I were to go to a website of a zoo, I can expect photos of caged animals. If I don't like to see those photos, I should not go to that site.

**Comment by joostdecock** [comment]
Regardless of the discussion, here's a 1uick example of implementation (from the top of my head):

```mjs
import { useState } from 'react'

export const Nsfw = ({ children }) => {
  const [show, setShow] = useState(false)
  
  return <div role="button" onClick={() => setShow(!show)} className={show ? '' : 'blur-lg'}>{children}</div>
}
```

Just to clarify what I mean by *why should not put stuff in globals.css if a tailwind class can do the trick*.

**Comment by HaasJona** [comment]
I am not sure moderation is a problem. I think the showcases that are posted are reviewed anyway, aren't they?

Deciding where to use the tag is of course a question. In my opinion, the main target would be showcases of underwear, swimwear and other revealing clothing that would be obviously more revealing than the pattern or the sewing documentation itself (which people would of course expect on the site).

Discussing the technical implementation is of course not needed if we decide we don't want this. I was mostly going the pure HTML/CSS route because I'm more comfortable there than with react.

**Comment by joostdecock** [comment]
I'm going to take an executive decision here and say we don't want this.

I hope @BenJamesBen and @HaasJona will accept my apologies for not questioning the validity of this request before they put the work in :see_no_evil: 

Closing
